src/annotator/test/config-test.js

```diff
@@ -1,86 +1,220 @@
+  });
+
+  context('when settings() returns a non-empty object', function() {
+    it('reads the setting into the returned options', function() {
+      // config() just blindly adds any key: value settings that settings()
+      // returns into the returns options object.
+      fakeSettings.returns({foo: 'bar'});
+
+      var options = config(fakeWindow());
+
+      assert.equal(options.foo, 'bar');
     });
   });
 
-  it('merges the config from hypothesisConfig()', function () {
-    var fakeWindow = Object.assign({}, fakeWindowBase, {
-      hypothesisConfig: function () {
-        return {foo: true};
-      },
+  context('when settings() throws an error', function() {
+    beforeEach(function() {
+      fakeSettings.throws();
+    });
+
+    it('catches the error', function() {
+      config(fakeWindow());
     });
-    assert.deepEqual(config(fakeWindow), {
-      app: 'app.html',
-      foo: true,
+
+    it('logs a warning', function() {
+      config(fakeWindow());
+
+      assert.called(console.warn);
     });
   });
 
-  it('merges the config from the "js-hypothesis-config" <script> tag', function () {
-    fakeScriptConfig = '{"annotations":"456"}';
-    assert.deepEqual(config(fakeWindowBase), {
-      app: 'app.html',
-      annotations: '456',
+  context("when there's a window.hypothesisConfig() function", function() {
+    it('reads arbitrary settings from hypothesisConfig() into options', function() {
+      var window_ = fakeWindow();
+      window_.hypothesisConfig = sinon.stub().returns({foo: 'bar'});
+
+      var options = config(window_);
+
+      assert.equal(options.foo, 'bar');
     });
+
+    specify('hypothesisConfig() settings override js-hypothesis-config ones', function() {
+      var window_ = fakeWindow();
+      window_.hypothesisConfig = sinon.stub().returns({
+        foo: 'fooFromHypothesisConfigFunc'});
+      fakeSettings.returns({foo: 'fooFromJSHypothesisConfigObj'});
+
+      var options = config(window_);
+
+      assert.equal(options.foo, 'fooFromHypothesisConfigFunc');
+    });
+
+    context('if hypothesisConfig() returns a non-object value', function() {
+      it("doesn't add anything into the options", function() {
+        var window_ = fakeWindow();
+        window_.hypothesisConfig = sinon.stub().returns(42);
+
+        var options = config(window_);
+
+        delete options.app; // We don't care about options.app for this test.
+        assert.deepEqual({}, options);
+      });
+    });
+  });
+
+  context("when window.hypothesisConfig() isn't a function", function() {
+    it('throws a TypeError', function() {
+      var window_ = fakeWindow();
+      window_.hypothesisConfig = 'notAFunction';
+
+      assert.throws(
+        function() { config(window_); }, TypeError,
+        'hypothesisConfig must be a function, see: https://h.readthedocs.io/en/latest/embedding.html'
+      );
+    });
+  });
+
+  describe('showHighlights', function() {
+    context("when it's true", function() {
```

Comment:
I ended up doing this:

```javascript
    [
      {
        name: 'changes `true` to `"always"`',
        in:   true,
        out:  'always',
      },
      {
        name: 'changes `false` to `"never"`',
        in:   false,
        out:  'never',
      },
      // It adds any arbitrary string value for showHighlights to the
      // returned options, unmodified.
      {
        name: 'passes arbitrary strings through unmodified',
        in:   'foo',
        out:  'foo',
      },
    ].forEach(function(test) {
      it(test.name, function() {
        fakeSettings.returns({showHighlights: test.in});

        var options = config(fakeWindow());

        assert.equal(options.showHighlights, test.out);
      });
    });
```

Wondering if there'd be any advantage to using `unroll` instead?